pinot/inference/adlala.py

```diff
@@ -1 +1,211 @@
+# MODULE CLASSES
+# =============================================================================
+class AdLaLa(torch.optim.Optimizer):
+    """ Adaptive Langevin-Langevin Integrator.
+
+    ref: `https://arxiv.org/abs/1908.11843`
+    
+    original problem-specific implementation:
+        `https://github.com/`
+        `TiffanyVlaar/`
+        `ThermodynamicParameterizationOfNNs/`
+        `blob/master/AdLaLa.ipynb`
+
+    Attributes
+    ----------
+    params : iterator
+        an iterator over a series of `torch.nn.Parameter` to be optimized.
+    h : float, default=0.25
+        learning rate.
+    gamma : float, dafault=0.1
+        friction parameter (for Langevin only).
+    sigma : float, default=0.01
+        driving noise amplitude (for Adaptive Langevin only).
+    epsilon: float, default=0.01
+        coupling coefficient (for Adaptive Langevin only).
+    tau: float, default=1e-4
+        temperature.
+    xi_init: float, default=1e-3
+        initial value for $xi$.
+
+    Methods
+    -------
+    step(closure): apply gradient.
+    
+
+    """
+
+    def __init__(
+            self,
+            params,
+            h=0.25,
+            gamma=0.1,
+            sigma=0.01,
+            epsilon=0.05,
+            tau=1e-4,
+            xi_init=1e-3,
+            partition='La'):
+    
+        defaults = dict(
+            h=torch.tensor(h),
+            tau=torch.tensor(tau),
+            gamma=torch.tensor(gamma),
+            sigma=torch.tensor(sigma),
+            epsilon=torch.tensor(epsilon),
+            xi_init=torch.tensor(xi_init),
+            partition='La')
+
+        super(AdLaLa, self).__init__(params, defaults)
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """ Performs a single optimization step.
+
+        Perform the steps in the following order:
+        
+        ```
+        A^1_(h/2) 
+        A^2_(h/2) 
+        C^1_(h/2) 
+        D^1_(h/2)
+        E^1_(h/2)
+        O^2_h
+        E^1_(h/2)
+        D^1_(h/2)
+        C^1_(h/2)
+        A^2_(h/2)
+        A^1_(h/2)
+        B^2_h
+        B^1_h
+        ```
+        where 1 denotes adaptive Langevin part
+        and 2 denotes Langevin part
+
+        the steps are:
+        ```
+        A: w := w + hp
+        B: p := p-h nabla L(w)
+        C: p := e^{-h * xi} p
+        D: p := p + sigma sqrt{h}R_n
+        E: xi := xi + h * epsilon [p^T p - N * tau]
+        O: p := c p + d R
+
+        
+        ```
+        """
+
+        # call closure if closure is specified
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        # loop through param groups with possible different parameter settings.
+        for group in self.param_groups:
+            for w in group['params']:
+                # NOTE: `w.grad == None` is different from `w.grad == 0.`
+                # as the later could be used to sample parameters without taking grads.
+                if w.grad == None:
```

Comment:
comparison to `None` should use `is`, `is not` rather than `==` https://www.python.org/dev/peps/pep-0008/#programming-recommendations